File: paytmlibrary/src/main/java/com/bigfun/tm/login/LoginSDK.java
package com.bigfun.tm.login;

import android.app.Activity;
import android.support.annotation.Keep;

import com.facebook.CallbackManager;

@Keep
public class LoginSDK {

    private static LoginSDK instance;
    private Activity activity;
    private LoginModel loginModel;

    private LoginSDK(Activity activity) {
        this.activity = activity;
        loginModel = new LoginModel(activity);
    }

    @Keep
    public static LoginSDK getInstance(Activity activity) {
        if (instance == null) {
            synchronized (LoginSDK.class) {
                if (instance == null) {
                    instance = new LoginSDK(activity);
                }
            }
        }
        return instance;
    }

    /**
     * facebook登录
     *
     * @param callbackManager
     * @param listener
     */
    @Keep
    public void facebookLogin(CallbackManager callbackManager, IFBLoginListener listener) {
        loginModel.facebookLogin(callbackManager, listener);
    }
}

File: paytmlibrary/src/main/java/com/bigfun/tm/login/LoginSdk.kt
package com.bigfun.tm.login

import android.app.Activity
import android.support.annotation.Keep
import com.facebook.CallbackManager

@Keep
class LoginSdk private constructor(private val activity: Activity) {

    private val loginModel by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
        LoginModel(activity)
    }

    @Keep
    companion object {

        @Volatile
        private var instance: LoginSdk? = null

        @JvmStatic
        fun getInstance(activity: Activity): LoginSdk {
            val i = instance
            if (i != null) {
                return i
            }

            return synchronized(this) {
                val i2 = instance
                if (i2 != null) {
                    i2
                } else {
                    val create = LoginSdk(activity)
                    instance = create
                    create
                }
            }
        }
    }

    /**
     * facebook登录
     */
    @Keep
    fun fbLogin(callbackManager: CallbackManager, listener: IFBLoginListener) {
        loginModel.facebookLogin(callbackManager, listener)
    }

    /**
     * google登录
     */
    @Keep
    private fun googleLogin(webClientId: String, requestCode: Int) {
//        loginModel.googleLogin(webClientId, requestCode)
    }
}